bitcoin/src/blockdata/script.rs

```diff
@@ -1222,6 +1232,101 @@ mod test {
         script = script.push_opcode(OP_CHECKSIG); comp.push(0xACu8); assert_eq!(&script[..], &comp[..]);
     }
 
+    #[test]
+    fn test_p2pk_pubkey_bytes() {
+        // happy path
+        // A valid key and a valid script should return Some and equal the key passed
+        let key_str = "0411db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5cb2e0eaddfb84ccf9744464f82e160bfa9b8b64f9d4c03f999b8643f656b412a3";
+        let key = PublicKey::from_str(key_str).unwrap();
+        let p2pk = Script::builder().push_key(&key).push_opcode(OP_CHECKSIG).into_script();
+        let actual = p2pk.p2pk_pubkey_bytes().unwrap();
+        assert_eq!(actual.to_vec(), key.to_bytes());
+
+        // invalid script checks
+        // A script without an OP_CHECKSIG should return None
+        let no_checksig = Script::builder().push_key(&key).into_script();
+        assert_eq!(no_checksig.p2pk_pubkey_bytes(), None);
+        // an empty script should return None
+        let empty_script = Script::builder().into_script();
+        assert!(empty_script.p2pk_pubkey_bytes().is_none());
+        // scripts with no key should return None
+        let no_push_bytes = Script::builder().push_opcode(OP_CHECKSIG).into_script();
+        assert!(no_push_bytes.p2pk_pubkey_bytes().is_none());
+        // scripts with a different op code other than OP_CHECKSIG should return None
+        let different_op_code = Script::builder().push_key(&key).push_opcode(OP_NOP).into_script();
+        assert!(different_op_code.p2pk_pubkey_bytes().is_none());
+
+        // invalid key check
+        // Should return None if the key is incorrect size (key is 63 bytes) 
+        let malformed_key = "21032e58afe51f9ed8ad3cc7897f634d881fdbe49816429ded8156bebd2ffd1";
+        let invalid_p2pk_script = Script::builder()
+            .push_slice(malformed_key.as_bytes())
+            .push_opcode(OP_CHECKSIG)
+            .into_script();
+        assert!(invalid_p2pk_script.p2pk_pubkey_bytes().is_none());
+        // invalid key should return Some, p2pk_pubkey_bytes doesn't care if the key is valid
+        let malformed_key = "21032e58afe51f9ed8ad3cc7897f634d881fdbe49816429ded8156bebd2ffd1ux";
+        let invalid_key_script = Script::builder()
+            .push_slice(malformed_key.as_bytes())
+            .push_opcode(OP_CHECKSIG)
+            .into_script();
+        assert!(invalid_key_script.p2pk_pubkey_bytes().is_some());
+
+        // compressed key tests
+        let compressed_key_str = "0311db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5c";
+        let key = PublicKey::from_str(compressed_key_str).unwrap();
+        let p2pk = Script::builder().push_key(&key).push_opcode(OP_CHECKSIG).into_script();
+        let actual = p2pk.p2pk_pubkey_bytes().unwrap();
+        assert_eq!(actual.to_vec(), key.to_bytes());
+    }
+
```

Comment:
Appreciate you adding unit tests mate. The following suggestion can totally be ignored if you so please. Adding it incase you are interested.

This test could be split up into many smaller tests, each with a single assertion. Each test can then be named with exactly what it does. FTR this idea is not fully followed in the library currently but I thought you might be interested. For example it could have been written as such
```rust

    #[test]
    fn p2pk_pubkey_bytes_valid_key_and_valid_script() {
        let key_str = "0411db93e1dcdb8a016b49840f8c53bc1eb68a382e97b1482ecad7b148a6909a5cb2e0eaddfb84ccf9744464f82e160bfa9b8b64f9d4c03f999b8643f656b412a3";
        let key = PublicKey::from_str(key_str).unwrap();
        let p2pk = Script::builder().push_key(&key).push_opcode(OP_CHECKSIG).into_script();
        let actual = p2pk.p2pk_pubkey_bytes().unwrap();
        assert_eq!(actual.to_vec(), key.to_bytes());
    }

    #[test]
    fn p2pk_pubkey_bytes_script_without_op_checksig_returns_none() {
        let no_checksig = Script::builder().push_key(&key).into_script();
        assert_eq!(no_checksig.p2pk_pubkey_bytes(), None);
    }

    #[test]
    fn p2pk_pubkey_bytes_empty_script_returns_none() {
        let empty_script = Script::builder().into_script();
        assert!(empty_script.p2pk_pubkey_bytes().is_none());
    }

    #[test]
    fn p2pk_pubkey_bytes_script_without_key_returns_none() {
        let no_push_bytes = Script::builder().push_opcode(OP_CHECKSIG).into_script();
        assert!(no_push_bytes.p2pk_pubkey_bytes().is_none());
    }

    #[test]
    fn p2pk_pubkey_bytes_script_with_opcode_other_than_opchecksig_returns_none() {
        let different_op_code = Script::builder().push_key(&key).push_opcode(OP_NOP).into_script();
        assert!(different_op_code.p2pk_pubkey_bytes().is_none());
    }

    #[test]
    fn p2pk_pubkey_bytes_incorrect_key_size_returns_none() {
        let malformed_key = "21032e58afe51f9ed8ad3cc7897f634d881fdbe49816429ded8156bebd2ffd1";
        let invalid_p2pk_script = Script::builder()
            .push_slice(malformed_key.as_bytes())
            .push_opcode(OP_CHECKSIG)
            .into_script();
        assert!(invalid_p2pk_script.p2pk_pubkey_bytes().is_none());
    }
```

I can't remember the exact source of this theory, I'm still learning about better testing also.

Reply:
Sure, I'll split 'em up

Reply:
Yes, there are three reasons to split the tests: the tests can run in parallel (faster), subset of the tests can be ran rather than forcing all of them (better development experience), when a test fails you can see the name of the test (this is only important if they are called via macros because line number is normally good enough).